File: src/com/kingmang/lazurite/parser/AST/Expressions/ContainerAccessExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.core.*;
import com.kingmang.lazurite.parser.AST.Accessible;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.*;
import com.kingmang.lazurite.runtime.Lzr.LzrArray;
import com.kingmang.lazurite.runtime.Lzr.LzrMap;
import com.kingmang.lazurite.runtime.Lzr.LzrString;
import lombok.Getter;


import java.util.List;


public final class ContainerAccessExpression implements Expression, Accessible {

    @Getter
    public final Expression root;
    @Getter
    public final List<Expression> indices;
    @Getter
    private boolean rootIsVariable;

    public ContainerAccessExpression(String variable, List<Expression> indices) {
        this(new VariableExpression(variable), indices);
    }

    public ContainerAccessExpression(Expression root, List<Expression> indices) {
        rootIsVariable = root instanceof VariableExpression;
        this.root = root;
        this.indices = indices;
    }


    @Override
    public Value eval() {
        return get();
    }
    
    @Override
    public Value get() {
        final Value container = getContainer();
        final Value lastIndex = lastIndex();
        switch (container.type()) {
            case Types.ARRAY:
                return ((LzrArray) container).get(lastIndex);

            case Types.MAP:
                return ((LzrMap) container).get(lastIndex);

            case Types.STRING:
                return ((LzrString) container).access(lastIndex);
                
            case Types.CLASS:
                return ((ClassInstanceValue) container).access(lastIndex);
                
            default:
                throw new LZRException("TypeExeption","Array or map expected. Got " + Types.typeToString(container.type()));
        }
    }

    @Override
    public Value set(Value value) {
        final Value container = getContainer();
        final Value lastIndex = lastIndex();
        switch (container.type()) {
            case Types.ARRAY:
                final int arrayIndex = lastIndex.asInt();
                ((LzrArray) container).set(arrayIndex, value);
                return value;

            case Types.MAP:
                ((LzrMap) container).set(lastIndex, value);
                return value;
                
            case Types.CLASS:
                ((ClassInstanceValue) container).set(lastIndex, value);
                return value;
                
            default:
                throw new LZRException("TypeExeption","Array or map expected. Got " + container.type());
        }
    }
    
    public Value getContainer() {
        Value container = root.eval();
        final int last = indices.size() - 1;
        for (int i = 0; i < last; i++) {
            final Value index = index(i);
            switch (container.type()) {
                case Types.ARRAY:
                    final int arrayIndex = index.asInt();
                    container = ((LzrArray) container).get(arrayIndex);
                    break;
                    
                case Types.MAP:
                    container = ((LzrMap) container).get(index);
                    break;
                    
                default:
                    throw new LZRException("TypeExeption","Array or map expected");
            }
        }
        return container;
    }
    
    public Value lastIndex() {
        return index(indices.size() - 1);
    }
    
    private Value index(int index) {
        return indices.get(index).eval();
    }
    
    public LzrMap consumeMap(Value value) {
        if (value.type() != Types.MAP) {
            throw new LZRException("TypeExeption","Map expected");
        }
        return (LzrMap) value;
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return root.toString() + indices;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/ContainerAccessExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.exceptions.FileInfo
import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.parser.ast.Accessible
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.ClassInstanceValue
import com.kingmang.lazurite.runtime.values.LzrArray
import com.kingmang.lazurite.runtime.values.LzrMap
import com.kingmang.lazurite.runtime.values.LzrString
import com.kingmang.lazurite.runtime.values.LzrValue

data class ContainerAccessExpression(
    val root: Expression,
    val indices: List<Expression>
) : Expression, Accessible {

    private var rootIsVariable : Boolean = false

    constructor(variable: String, indices: List<Expression>, file: FileInfo) : this(VariableExpression(variable, file), indices)
    {
        rootIsVariable = root is VariableExpression;
    }
    override fun eval(): LzrValue =
        this.get()

    override fun get(): LzrValue {
        val container = this.container
        val lastIndex = this.lastIndex()
        return when (container.type()) {
            Types.ARRAY -> (container as LzrArray)[lastIndex]
            Types.MAP -> (container as LzrMap)[lastIndex]!!
            Types.STRING -> (container as LzrString).access(lastIndex)
            Types.CLASS -> (container as ClassInstanceValue).access(lastIndex)
            else -> throw LzrException("TypeException", "Array or map expected. Got " + Types.typeToString(container.type()))
        }
    }

    override fun set(value: LzrValue): LzrValue {
        val container = this.container
        val lastIndex = lastIndex()
        when (container.type()) {
            Types.ARRAY -> (container as LzrArray)[ lastIndex.asInt()] = value
            Types.MAP -> (container as LzrMap)[lastIndex] = value
            Types.CLASS -> (container as ClassInstanceValue)[lastIndex] = value
            else -> throw LzrException("TypeException", "Array or map expected. Got " + container.type())
        }
        return value
    }

    val container: LzrValue
        get() {
            var container = this.root.eval()
            val last = this.indices.size - 1
            for (i in 0 until last) {
                val index = index(i)
                container = when (container.type()) {
                    Types.ARRAY -> (container as LzrArray)[index.asInt()]
                    Types.MAP -> (container as LzrMap)[index]!!
                    Types.CLASS -> (container as ClassInstanceValue)[index]
                    else -> throw LzrException("TypeException", "Array or map expected")
                }
            }
            return container
        }

    private fun lastIndex(): LzrValue =
        index(this.indices.size - 1)

    private fun index(index: Int): LzrValue =
        this.indices[index].eval()

    fun consumeMap(value: LzrValue): LzrMap {
        if (value.type() != Types.MAP)
            throw LzrException("TypeException", "Map expected")
        return value as LzrMap
    }
    fun rootIsVariable() : Boolean {
        return rootIsVariable;
    }
    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R =
        visitor.visit(this, t)

    override fun toString(): String =
        this.root.toString() + this.indices
}
